### PR TITLE
Bloom: add optional ping-pong implementation.

### PR DESCRIPTION
**Pull request by prideout**
For now, this is only active for WebGL but we will likely activate it for Intel GPU's in an upcoming PR.

#2338

### CODE REVIEW

**Comment by prideout** [comment]
To recap, some browsers do not allow reading from a mip while writing to a different mip. In the OpenGL ES 3.0 specification this is covered in section 4.4.3, "Feedback Loops Between Textures and the Framebuffer".

Chromium bug at https://bugs.chromium.org/p/chromium/issues/detail?id=1066201

File: filament/src/PostProcessManager.cpp [hunk 1]
```diff
@@ -1308,27 +1308,50 @@ FrameGraphId<FrameGraphTexture> PostProcessManager::bloomPass(FrameGraph& fg,
         });
     }
 
+    // TODO: this is a feature flag that should be passed in from above.
```
Comment (bejado) [inline filament/src/PostProcessManager.cpp:1311]: I would maybe add a comment here explaining what "feedback" means and why it's necessary to disable it in certain cases.

**Review by bejado — APPROVED** [review]

File: filament/src/PostProcessManager.cpp (hunk 1 above)
Comment (prideout) [reply filament/src/PostProcessManager.cpp:1311]: Good catch, this term is from the GL spec, I amended the comment by fleshing out this comment block.

File: filament/src/PostProcessManager.cpp [hunk 2]
```diff
@@ -1308,27 +1308,53 @@ FrameGraphId<FrameGraphTexture> PostProcessManager::bloomPass(FrameGraph& fg,
...
+                if (disableFeedback) {
+                    data.stage = builder.createTexture("Bloom Stage Texture", {
+                            .width = width,
+                            .height = height,
+                            .levels = bloomOptions.levels,
+                            .format = outFormat
+                    });
+                    data.stage = builder.write(builder.sample(data.stage));
+                } else {
+                    data.stage = builder.write(builder.sample(data.out));
```
Comment (pixelflinger) [inline filament/src/PostProcessManager.cpp:1350]: one thing that makes me nervous here is that `data.out` becomes invalid after calling `sample`. So it's dangerous to have it around. At least we need a comment, and maybe reset it to 0.

File: filament/src/PostProcessManager.cpp [hunk 3]
```diff
@@ -1308,27 +1308,53 @@ FrameGraphId<FrameGraphTexture> PostProcessManager::bloomPass(FrameGraph& fg,
...
+                }
+
                 for (size_t i = 0; i < bloomOptions.levels; i++) {
-                    data.outRT[i] = builder.createRenderTarget("Bloom target", {
+                    data.outRT[i] = builder.createRenderTarget("Bloom Out Target", {
                             .attachments = {{ data.out, uint8_t(i) }} });
```
Comment (pixelflinger) [inline filament/src/PostProcessManager.cpp:1355]: I dont know why it doesn't fail -- but data.out is invalid here in the normal case (see my comment above)

File: filament/src/PostProcessManager.cpp [hunk 4]
```diff
@@ -1353,45 +1380,53 @@ FrameGraphId<FrameGraphTexture> PostProcessManager::bloomPass(FrameGraph& fg,
...
     // upsample phase
     auto& bloomUpsamplePass = fg.addPass<BloomPassData>("Bloom Upsample",
             [&](FrameGraph::Builder& builder, auto& data) {
-                data.in = builder.sample(input);
-                data.out = builder.write(input);
+                data.out = builder.write(builder.sample(output));
+                if (disableFeedback) {
+                    data.stage = builder.write(builder.sample(stage));
+                } else {
+                    data.stage = builder.write(data.out);
```
Comment (pixelflinger) [inline filament/src/PostProcessManager.cpp:1416]: same problem here

File: filament/src/PostProcessManager.cpp [hunk 5]
```diff
@@ -1353,45 +1380,53 @@ FrameGraphId<FrameGraphTexture> PostProcessManager::bloomPass(FrameGraph& fg,
                 mi->setParameter("invHighlight", std::isinf(bloomOptions.highlight) ? 0.0f : 1.0f / bloomOptions.highlight);
 
                 for (size_t i = 0; i < bloomOptions.levels; i++) {
-                    auto hwOutRT = resources.get(data.outRT[i]);
+                    const bool parity = (i % 2) == 0;
```
Comment (pixelflinger) [inline filament/src/PostProcessManager.cpp:1383]: does this code work if the parity if `bloomOptions.levels` changes?

File: web/filament-js/extensions.js [hunk 6]
```diff
@@ -251,11 +251,12 @@ Filament.loadClassExtensions = function() {
 
     /// setBloomOptions ::method::
     /// overrides ::argument:: Dictionary with one or more of the following properties: \
-    /// dirtStrength, strength, resolution, anomorphism, levels, blendMode, threshold, enabled.
+    /// enabled, strength, resolution, anomorphism, levels, blendMode, threshold, highlight.
     /// NOTE: dirt texture is not yet supported in the JavaScript API.
     Filament.View.prototype.setBloomOptions = function(overrides) {
         const options = {
             dirtStrength: 0.2,
+            highlight: 1000.0,
```
Comment (pixelflinger) [inline web/filament-js/extensions.js:259]: should be in another PR?

**Review by pixelflinger — CHANGES_REQUESTED** [review]
That makes me very nervous!  My preference would be to disable bloom entirely on these GPUs.

File: web/filament-js/extensions.js (hunk 6 above)
Comment (prideout) [reply web/filament-js/extensions.js:259]: At least it's a separate commit but yeah it should've made a sep PR too.

**Comment by romainguy** [comment]
Mathias, that would mean disabling bloom entirely on Intel GPUs (so most laptops basically) and in WebGL.

File: filament/src/PostProcessManager.cpp (hunk 5 above)
Comment (prideout) [reply filament/src/PostProcessManager.cpp:1383]: Yes, I had to think about this: the parity of the downsample pass is straightforward but the parity of the upsample is adjusted accordingly.

File: filament/src/PostProcessManager.cpp (hunk 2 above)
Comment (prideout) [reply filament/src/PostProcessManager.cpp:1350]: The current code is trying to make it so that "stage" is just an alias for "out".  Would is make sense to add an alias method to the builder for this?

Another idea is leaving `bloomPass` exactly as it is, and instead simply create a brand new method called `bloomPingPong`.  There will be some code duplication but maybe it will be safer.

@pixelflinger WDYT?

File: web/filament-js/extensions.js (hunk 6 above)
Comment (prideout) [reply web/filament-js/extensions.js:259]: Gah, you're right, there is some cross-over between the two CL's.

**Comment by prideout** [comment]
Discarding this in favor of a new cleaner approach, stay tuned.
